source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -76,65 +143,75 @@ Http::FilterHeadersStatus CacheFilter::encodeHeaders(Http::ResponseHeaderMap& he
+void CacheFilter::getTrailers() {
+  ASSERT(lookup_, "CacheFilter is trying to call getTrailers with no LookupContext");
+  ASSERT(response_has_trailers_, "No reason to call getTrailers when there's no trailers to get.");
+  lookup_->getTrailers(
+      [this](Http::ResponseTrailerMapPtr&& trailers) { onTrailers(std::move(trailers)); });
+}
+
 void CacheFilter::onHeaders(LookupResult&& result) {
   // TODO(yosrym93): Handle request only-if-cached directive
   switch (result.cache_entry_status_) {
   case CacheEntryStatus::FoundNotModified:
   case CacheEntryStatus::UnsatisfiableRange:
     NOT_IMPLEMENTED_GCOVR_EXCL_LINE; // We don't yet return or support these codes.
   case CacheEntryStatus::RequiresValidation:
-    // Cache entries that require validation are treated as unusable entries
-    // until validation is implemented
-    // TODO(yosrym93): Implement response validation
+    // If a cache entry requires validation inject validation headers in the request and let it
+    // pass through as if no cache entry was found. If the cache entry was valid the response
+    // status should be 304 (unmodified), and the cache entry will be injected in the response body
+    validating_cache_entry_ = true;
+    lookup_result_ = std::make_unique<LookupResult>(std::move(result));
+    injectValidationHeaders();
   case CacheEntryStatus::Unusable:
-    if (state_ == GetHeadersState::FinishedGetHeadersCall) {
+    if (get_headers_state_ == GetHeadersState::FinishedGetHeadersCall) {
       // decodeHeader returned Http::FilterHeadersStatus::StopAllIterationAndWatermark--restart it
       decoder_callbacks_->continueDecoding();
     } else {
       // decodeHeader hasn't yet returned--tell it to return Http::FilterHeadersStatus::Continue.
-      state_ = GetHeadersState::GetHeadersResultUnusable;
+      get_headers_state_ = GetHeadersState::GetHeadersResultUnusable;
     }
-    return;
+    break;
   case CacheEntryStatus::Ok:
-    response_has_trailers_ = result.has_trailers_;
-    const bool end_stream = (result.content_length_ == 0 && !response_has_trailers_);
-    // TODO(toddmgreer): Calculate age per https://httpwg.org/specs/rfc7234.html#age.calculations
-    result.headers_->addReferenceKey(Http::Headers::get().Age, 0);
-    decoder_callbacks_->streamInfo().setResponseFlag(
-        StreamInfo::ResponseFlag::ResponseFromCacheFilter);
-    decoder_callbacks_->streamInfo().setResponseCodeDetails(
-        CacheResponseCodeDetails::get().ResponseFromCacheFilter);
-    decoder_callbacks_->encodeHeaders(std::move(result.headers_), end_stream);
-    if (end_stream) {
-      return;
-    }
-    if (result.content_length_ > 0) {
-      remaining_body_.emplace_back(0, result.content_length_);
-      getBody();
-    } else {
-      lookup_->getTrailers(
-          [this](Http::ResponseTrailerMapPtr&& trailers) { onTrailers(std::move(trailers)); });
-    }
+    lookup_result_ = std::make_unique<LookupResult>(std::move(result));
+    encodeCachedResponse();
```

Comment:
The code removed from `case CacheEntryStatus::Ok:` is refactored into `encodeCachedResponse()` with some modifications so that it works for encoding cached responses in both decoding and encoding scenarios.

decoding scenario -> fresh cache hit found when decoding
encoding scenario -> stale cache hit found when decoding, successfully validated when encoding